### PR TITLE
Collection curation page

### LINKED ISSUE #725: Collection facet/filter
Implement a "collection" facet to filter things that are in the chosen collection.

#724

### PR DESCRIPTION
**Pull request by DaanVanVugt**
- [x] translations
- [x] controller test
~~curation reminder mail (will do in other PR)~~
- [x] add collection membership to SOLR
- [x] triggers to update SOLR indexing when changing collection title, destroying collection or adding elements

Fixes #725

### CODE REVIEW

File: config/routes.rb
Comment (DaanVanVugt) [inline config/routes.rb:94]: this return may cause trouble, should be `next`. I notice we don't have rubocop or anything like that on this repo, should we set that up?

File: app/models/material.rb
Comment (fbacall) [inline app/models/material.rb:70]: Could use `.pluck(:title)` here for a slightly simpler DB query

File: app/views/collections/show.html.erb
Comment (fbacall) [inline app/views/collections/show.html.erb:18]: I would also add a feature flag to `tess.yml` to enable/disable the collection curation feature.

File: app/controllers/collections_controller.rb
Comment (fbacall) [inline app/controllers/collections_controller.rb:53]: You may also want to add the `.from_verified_users` scope to avoid spam.

File: config/routes.rb
Comment (fbacall) [reply config/routes.rb:94]: and generally I don't really like disabling features at the routes level, IMO it should be via a `before_action` in the controllers. I'll probably create an issue relating to this.

**Comment by fbacall** [comment]
@DaanVanVugt dunno if you saw my review comments.

None of them are blockers really, so happy to merge if you are not working on this currently.

**Comment by DaanVanVugt** [comment]
I can continue work on this in November, so feel free to leave it for now.

Op di 25 okt. 2022 11:37 schreef Finn Bacall ***@***.***>:

**Comment by DaanVanVugt** [comment]
@fbacall I have implemented your review comments, except for the feature flag. My editor has also automatically made some changes rubocop suggests, so sorry for the spam because of that.

File: app/views/collections/show.html.erb
Comment (DaanVanVugt) [reply app/views/collections/show.html.erb:18]: That's good of course. I have added the check in the policy.

**Comment by fbacall** [comment]
> My editor has also automatically made some changes rubocop suggests, so sorry for the spam because of that.
No worries. I wonder if we should do a complete `rubocup -a` pass on the repository...

**Comment by fbacall** [comment]
Is this good to merge now?

**Comment by DaanVanVugt** [comment]
> > My editor has also automatically made some changes rubocop suggests, so sorry for the spam because of that.
> 
> No worries. I wonder if we should do a complete `rubocup -a` pass on the repository...

yes, I think that would be good to do. Can ignore that commit using `.git-ignore-revs` (see https://www.stefanjudis.com/today-i-learned/how-to-exclude-commits-from-git-blame/) also on github

**Comment by DaanVanVugt** [comment]
> Is this good to merge now?

perhaps one manual end-to-end test makes sense still, after the merge. I'll see if I have time to do that tonight.

**Comment by DaanVanVugt** [comment]
I have tested it myself, and fixed a few small things still. Only thing to still look at perhaps is this Event time handling concern added in the last commit. Then, if it passes CI still, this is fine to merge in my opinion.

**Comment by DaanVanVugt** [comment]
interestingly the CI here fails, but works locally, due to the i18n.raise_on_missing which is not yet in this branch (but will be in the merged version, as it is in master). I think it's fixed now.

**Comment by DaanVanVugt** [comment]
After trying it out on staging.taxila.nl I think it will be good to add also a lower limit of the current date by default, to avoid loading a ridiculous amount of events. So not Time.at(0), but Time.zone.today

**Comment by fbacall** [comment]
I merged this https://github.com/ElixirTeSS/TeSS/pull/760 which will probably break this work (but hopefully makes things simpler?)

File: app/controllers/collections_controller.rb
Comment (DaanVanVugt) [inline app/controllers/collections_controller.rb:121]: probably this can really only be removed if we choose to have a single page for curating events and materials at the same time

File: app/controllers/collections_controller.rb
Comment (fbacall) [inline app/controllers/collections_controller.rb:161]: Could this be simplified by just doing e.g. `@collection.event_ids = (selected + existing IDs?)`

Comment (DaanVanVugt) [reply app/controllers/collections_controller.rb:161]: something like that, but not completely (we need to filter for the reviewed item ids only)

Comment (DaanVanVugt) [reply app/controllers/collections_controller.rb]: that approach would also load _all_ existing entries from the database, which is not necessary. I rewrote to use bulk inserts and deletes.

File: app/controllers/collections_controller.rb
Comment (fbacall) [inline app/controllers/collections_controller.rb:176]: I still think this could simplified.

What about:
```ruby
method = "#{item_class.model_name.singular}_ids"
remove = params[:reviewed_item_ids]&.map(&:to_i)
keep_or_add = params[:item_ids]&.map(&:to_i)
@collection.send("#{method}=", (@collection.send(method) - remove) |  keep_or_add))
```

That should set the order, perform indexing etc. without having to do it manually

File: app/models/concerns/curation_queue.rb
Comment (fbacall) [inline app/models/concerns/curation_queue.rb:34]: Is this needed? I think `user` is a required field on all curated types

File: app/models/concerns/curation_queue.rb
Comment (fbacall) [inline app/models/concerns/curation_queue.rb:39]: As above

File: app/models/event.rb
Comment (fbacall) [inline app/models/event.rb:91]: `.pluck(:title)` would be faster

File: app/models/material.rb
Comment (fbacall) [reply app/models/material.rb:70]: as above

File: config/routes.rb
Comment (fbacall) [inline config/routes.rb:78]: Better to check feature in controller: https://github.com/ElixirTeSS/TeSS/issues/729

File: app/models/concerns/curation_queue.rb
Comment (DaanVanVugt) [reply app/models/concerns/curation_queue.rb:34]: I ran into issues locally where there was no user set, which caused a crash. I think in general it is nice to be a little resilient to data quality issues.

File: app/models/event.rb
Comment (DaanVanVugt) [reply app/models/event.rb:91]: only if the collections are not already in memory (which they could be). But I'll change it to pluck, and a performance analysis can be done at a later stage.

File: app/controllers/collections_controller.rb
Comment (DaanVanVugt) [reply app/controllers/collections_controller.rb]: I have written a new version, which uses subqueries to only select the items which did not exist yet, or which already existed (for add / rm respectively), using a query like:

```sql
SELECT "events".*
FROM "events"
WHERE "events"."id" = $1
  AND "events"."id" != $2
  AND (EXISTS
         (SELECT 1
          FROM "events"
          INNER JOIN "collection_items" ON "events"."id" = "collection_items"."resource_id"
          WHERE "collection_items"."collection_id" = 1
            AND "collection_items"."resource_type" = 'Event'
          ORDER BY "collection_items"."order" ASC))
```

Comment (fbacall) [reply app/controllers/collections_controller.rb]: > Using item_ids instantiates all of the models (even the unchanged ones), which may be problematic for large collections. See collection_singular_ids= on https://apidock.com/rails/ActiveRecord/Associations/ClassMethods/has_many

Interesting! I would not have expected that.

File: app/models/concerns/curation_queue.rb
Comment (fbacall) [reply app/models/concerns/curation_queue.rb:34]: I think it would be better to address why the user was nil at that point, because I imagine a nil user will cause numerous other issues, and should be enforced by the association/validation

Comment (DaanVanVugt) [reply app/models/concerns/curation_queue.rb:34]: Alright, will remove these safeties.

**Review by fbacall — APPROVED** [review]

**Pull request merged**
